Source/API/Extras/CBLIncrementalStore.m

```diff
@@ -761,239 +634,797 @@ - (void) _initializeViews
+        NSMutableArray *result = [NSMutableArray arrayWithCapacity:[objects count]];
+        for (NSManagedObject *object in objects) {
+            NSDictionary *properties =
+            [self _couchbaseLiteRepresentationOfManagedObject:object withCouchbaseLiteID:YES];
+
+            if (propertiesToFetch) {
+                [result addObject:[properties dictionaryWithValuesForKeys:propertiesToFetch]];
+            } else {
+                [result addObject:properties];
+            }
+        }
+        return result;
+    } else {
+        if (outError) {
+            NSString *desc = [NSString stringWithFormat:@"Unsupported fetch request result type: %d", (int)type];
+            *outError = [NSError errorWithDomain:kCBLISErrorDomain
+                                            code:CBLIncrementalStoreErrorUnsupportedFetchRequestResultType
+                                        userInfo:@{ NSLocalizedFailureReasonErrorKey: desc }];
         }
+        return nil;
     }
-              version:@"1.0"];
-    
-    [self _setViewName:viewName forFetchingProperty:propertyName fromEntity:entityName];
 }
 
-#pragma mark - Querying
+#pragma mark - Query (QueryBuilder)
+
+- (NSArray *)fetchByUsingQueryBuilderWithPredicate:(NSPredicate *)predicate
+                               withOriginalRequest:(NSFetchRequest *)request
+                                  withTemplateVars:(NSDictionary *)templateVars
+                                       withContext:(NSManagedObjectContext *)context
+                                          outError:(NSError **)outError {
+    NSPredicate *typePredicate = [NSPredicate predicateWithFormat:@"%K == %@", kCBLISTypeKey,
+                                  request.entityName];
+
+    NSPredicate *compoundPredicate = [NSCompoundPredicate
+                                      andPredicateWithSubpredicates:@[typePredicate, predicate]];
+
+    CBLQueryBuilder *builder = [self cachedQueryBuilderForPredicate:compoundPredicate
+                                                    sortDescriptors:request.sortDescriptors];
+    if (!builder) {
+        builder = [[CBLQueryBuilder alloc] initWithDatabase:self.database
+                                                     select:nil
+                                             wherePredicate:compoundPredicate
+                                                    orderBy:request.sortDescriptors
+                                                      error:outError];
+    }
 
-/** Queries the database by a given fetch request. Checks the cache for the result first. */
-- (NSArray*) _queryObjectsOfEntity:(NSEntityDescription*)entity byFetchRequest:(NSFetchRequest*)fetch inContext:(NSManagedObjectContext*)context error:(NSError**)outError
-{
-    id cached = [self _cachedQueryResultsForEntity:entity.name predicate:fetch.predicate];
-    if (cached) {
-        return cached;
+    if (!builder)
+        return nil;
+
+    [self cacheQueryBuilder:builder
```

Comment:
Don't you only need to cache the builder if it's new? You can move lines 774-779 inside the if() block above.